### DISCUSSION TITLE
How can I tell when a client is offline?

### DISCUSSION THREAD

**Discussion by edalzell** [qruto/laravel-wave#45] [Q&A] [answered]
I need to do “stuff” when a user goes offline (close window), in a chat. Is there a way to detect that via Wave somehow?

I’m trying to use the Echo Presence stuff but when the use closes the browser I want to mark them offline.

anyone done this?

**Comment by slavarazum** [comment]
The "leave" event for presence channels is triggered by the [`beforeunload` window event](https://github.com/qruto/laravel-wave-client/blob/main/src/echo-broadcaster/wave-presence-channel.ts#L16). This event ensures that the channel is unsubscribed when the window is closed.

**Reply by edalzell** [reply]
that event is not reliable (I've found), but I do see that when the ping fails a "leave" event is triggered. Having some difficulties with Livewire/Echo events, not sure where the issue is.

**Reply by slavarazum** [reply]
The PHP script cannot determine when a connection has been aborted. It can only verify if the connection is still active when a new event occurs. Therefore, if the `beforeunload` event is not triggered to send the `leave` event when the user closes the window, the `ping` event will double-check and discard outdated connections. However, when the user closes a browser window tab, the `beforeunload` event is fired as expected.

**Reply by edalzell** [reply]
Any chance I can hire you for an hour to help me through this issue? trying to launch a SaaS platform and this is the last major unresolved issue. Bet you can solve it quickly.

**Comment by slavarazum** [comment, marked as answer]
You can manage a user's online status using [presence channels](https://laravel.com/docs/11.x/broadcasting#presence-channels).
